test(conformance): Enable smoke test in ipv6 cluster

sayboras opened the pull request:
This PR is to enable  conformance test for ipv6-only cluster.

Due to the lack of IPv6 support in github action, two tests related to 
outgoing traffic (1.1.1.1 and www.google.com) are skipped now.

I take this chance to standardise env variable in IPv4 smoke
test.

Relates #12201

Signed-off-by: Tam Mach <sayboras@yahoo.com>

**_Note:_** This PR is also having new test cases added in https://github.com/cilium/cilium/pull/12599.
I will rebase and add auto generated steps later, just mark this 
PR ready for early feedback

```release-note
test(conformance): Enable ipv6 smoke test in ipv6 cluster
```

brb commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,91 @@
+          helm upgrade -i cilium ./install/kubernetes/cilium \
+            --namespace kube-system \
+            --set global.nodeinit.enabled=true \
+            --set global.kubeProxyReplacement=partial \
```
You can use `--set global.kubeProxyReplacement=strict`. This will allow us to to remove the `nodePort`, `externalIPs`, `hostServices` `hostPort` settings.

sayboras replied:
oh nice, I didn't know about this _strict_ option. Thanks

rolinh replied:
Mmh, I don't think that `kubeProxyReplacement=strict` plays well with `kind`. I'd stick to `partial` for this one.

joestringer replied:
@Rolinh Is it kind or is it the underlying kernel that kind is running on? I'd +1 on `strict` if it works in CI. Not sure what kernel we get there.

rolinh replied:
@joestringer Not sure what the real issue is but if I enable `strict` on a `kind` cluster on my machine (with kernel `5.7.10`) I run into all sort of issues such as coredns pods not getting into ready-state, cilium pod in crashloop trying to reach the kube api server, etc.

brb commented on examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml:
```diff
@@ -0,0 +1,1338 @@
+    name: echo-b-host
+---
+apiVersion: v1
+kind: Service
```
Once https://github.com/cilium/cilium/issues/9776 has been fixed, we can set the `ipFamily`.

aanm replied:
It does not hurt to set it before #9776 is fixed.

sayboras replied:
sure, I tried to achieve this based on flag ipFamly  from cue command, but the code is quite ugly, maybe I need to read cuelang docs a little bit more. Let me do it as follow up.

sayboras committed:
test(conformance): Enable smoke test in ipv6 cluster

To enable  conformance test for ipv6-only cluster.

Due the lack of ipv6 support in github actions, two related tests
(1.1.1.1 and www.google.com FQDN) are skipped now.
Add auto-generate scritpt for connectivity-check-internal.yaml

Update generation Makefile target for connectivity check

- Use ">>" instead of ">" to avoid header got overriden
- Add $(ALL_TARGET) into .PHONY list
- Use $@ annotation instead of repated variable to avoid potential mistake

Signed-off-by: Tam Mach <sayboras@yahoo.com>

tgraf commented on .github/workflows/smoke-test.yaml:
```diff
@@ -60,49 +62,43 @@ jobs:
         run: |
           kubectl apply -f ${{ matrix.target.template }}
-          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=5m
+          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=${{ env.TIMEOUT }}
```
Can you split this into a separate commit and explain why the default timeout is changed from 5m to 2m?

sayboras replied:
makes sense :+1:

tgraf commented on examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml:
```diff
@@ -0,0 +1,1338 @@
+# Automatically generated by Makefile. DO NOT EDIT
```
This doesn't seem to be automatically generated but a copy of the generated yaml that is then modified. This means that this will get oudated quickly as it will not receive the same fixes. Can we instead automatically generate this from the IPv4 version?

sayboras replied:
agreed, I wait for #12599 then will rebase and add script to generate this.

sayboras replied:
added cue script to generate this file :)

tgraf reviewed (changes requested):
Looks awesome, two small requests

fristonio commented on .github/kind-config-ipv6.yaml:
```diff
@@ -0,0 +1,18 @@
+  ipFamily: ipv6
+  disableDefaultCNI: true
+  podSubnet: "fd00:10:244::/64"
+  serviceSubnet: "fd00:10:96::/112"
```
Needs a new line.

fristonio commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,88 @@
+            --set global.tunnel=disabled \
+            --set global.autoDirectNodeRoutes=true
+
+          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=5m
```
I think you missed using `env.TIMEOUT` here.

sayboras replied:
good catch :+1:

fristonio commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,88 @@
+            --set global.autoDirectNodeRoutes=true
+
+          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=5m
+          # To make sure that cilium CRD is available (default timeout is 5m)
```
Are we configuring the default timeout to be 5m or 2m?

sayboras replied:
oh thanks, the comment seems outdated as there was recent change in operator CRD. Let me clean it up.

fristonio commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,88 @@
+              kubectl describe pod $p
+              echo ---------------------------------------------------------------------------------
+              kubectl logs $p --since=${{ env.LOG_TIME }}
+          done
```
Needs a new line at the end.

fristonio reviewed (changes requested):
Overall looks great :rocket:

joestringer commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,88 @@
+          kubectl apply -f ${{ env.CONFORMANCE_TEMPLATE }}
+          kubectl wait --for=condition=Available --all deployment --timeout=${{ env.TIMEOUT }}
+
+      - name: Dump cilium related logs and events
```
:thinking: How do we keep these consistent across different checks?

(Worth also noting that both https://github.com/cilium/cilium/pull/12689 and https://github.com/cilium/cilium/pull/12599 will likely extend the existing GH output we gather for debugging)

sayboras replied:
@joestringer good point, what do think about add one more label (e.g. category) so that we can just filter based on label ?

sayboras replied:
it could be based on the logical group in generation script.

joestringer replied:
I'm not sure whether I follow. How would we configure / use the label?

My concern is that this is a brand new YAML for IPv6 and it may diverge from the IPv4 version. Maybe my concern is overblown if there's just two versions of the YAMLs, we can just ensure that we update both. I'd just want to make sure that any debuggability improvements we make will apply to both versions of the check.

joestringer replied:
Hmm, actually I just hit this type of problem with #12599 where I *thought* #12689 would help by printing the info I need when it fails, however I updated the "dump logs" section for the wrong check so I get no more information than before. As always, easy fix but also easy to miss on initial submission :-)

sayboras replied:
thanks, I just rebased with your changes to make it standardized.

nebril commented on examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml:
```diff
@@ -0,0 +1,1338 @@
+            - port: "5353"
+              protocol: UDP
+---
+#apiVersion: apps/v1
```
Can we remove this leftover?

sayboras replied:
thanks, done :)

nebril reviewed (commented):
Cool stuff! Let's make the yaml generated and ship it!

sayboras commented on .github/workflows/smoke-test-ipv6.yaml:
```diff
@@ -0,0 +1,87 @@
+            --set global.externalIPs.enabled=true \
+            --set global.nodePort.enabled=true \
+            --set global.hostPort.enabled=true \
+            --set config.bpfMasquerade=false \
```
I added bpfMasquerade flag to make it in sync with ipv4 conformance test. Thanks to @joestringer

fristonio commented on examples/kubernetes/connectivity-check/.gitignore:
```diff
@@ -0,0 +1 @@
+servers.yaml
```
Thanks for this.

fristonio commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -30,6 +32,10 @@ $(DEFAULT_OUT): $(SRC)
 
+$(IPv6_ONLY_OUT): $(SRC)
+	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(IPv6_ONLY_OUT)
+	$(QUIET)$(CUE) -t ipFamily=IPv6 dump > $(IPv6_ONLY_OUT)
```
Sorry, I missed out on the review of Joe's PR on Cue usage in connectivity-checks. But I think this line should append the cue dump `$(QUIET)$(CUE) -t ipFamily=IPv6 dump >> $(IPv6_ONLY_OUT)` to keep the autogenerated header on the top of the file, right?

sayboras replied:
this is very good catch, thanks a lot :100: 

Let me take this chance to update for other targets as well.

fristonio replied:
That would be great. :vulcan_salute:

sayboras commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -28,24 +30,28 @@ all: $(ALL_TARGETS)
 $(HOSTPORT_OUT): $(SRC)
 	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(HOSTPORT_OUT)
-	$(QUIET)$(CUE) -t component=all dump > $(HOSTPORT_OUT)
+	$(QUIET)$(CUE) -t component=all dump >> $@
```
For reviewer, there was a mixed usage of $@ and $(OUT_FILE) in this Makefile. As I am changing this file, I just change all to $@ for standardization.

$@
The file name of the target of the rule. If the target is an archive member, then ‘$@’ is the name of the archive file. In a pattern rule that has multiple targets (see Introduction to Pattern Rules), ‘$@’ is the name of whichever target caused the rule’s recipe to be run.

fristonio replied:
While we are at this, I am not sure why there is an inconsistent usage of `@` and `$(QUIET)` in the Makefile. Is there a particular reason we are using a mix of `@` and `$(QUIET)` here @joestringer ?

sayboras replied:
I think we can control the baviour based of `$(QUIET)` based on `$(V)` (e.g. export V=0), meanwhile, `@` behavior is unchangable.

fristonio replied:
But I still don't understand if a mix of these makes sense here. :sweat_smile:
Anyways the changes look good to me overall and this is not a problem.

joestringer replied:
Re: inconsistent usage, for some cases I figure that when you want to debug or understand the process occurring, you would want to see the commands being run. Typically for `echo` commands (especially when printing output to the terminal), this is not useful because you end up with duplicate printing: Once in the command output and once in the stdout.

Though that said I almost exclusively turn $(QUIET) on so I forget what the build looks like without it :-)

joestringer commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -5,6 +5,8 @@ include ../../../Makefile.defs
 
 DEFAULT_OUT := connectivity-check.yaml
+# Same as connectivity-check.yaml, but without 1.1.1.1 and google.com
+IPv6_ONLY_OUT := connectivity-check-ipv6.yaml
```
nit: I wonder if it's worth calling this `internal` or `cluster` or similar given that the actual implementation is not changing to guarantee ipv4 / ipv6, it's just turning on/off external checks.

sayboras replied:
This is really better naming, let me make changes for this.

sayboras replied:
I have made changes, please help to take another :eyes: whenever you have time. Thanks :)

joestringer reviewed (commented):
CUE usage looks good, although I think it's a little misleading to call this "ipv6" at the connectivity check layer since all it does is exclude the external checks. If IPv4 is available, it will still use IPv4.

I think if we were to specifically call out IPv4/IPv6, we should consider passing IP family parameters to the curl commands to ensure IPv6 is used. That said I don't think this is necessary for the goals of this PR; in the context of the IPv6 GH smoke test if we're disabling IPv4 then it should only perform IPv6 checks (because any IPv4 connections will be dropped).

I didn't look closely at the smoke test YAML, I figured others have more experience with what that should look like. Feel free to re-request my review if you are not getting sufficient eyes on that area.

joestringer commented on examples/kubernetes/connectivity-check/main_tool.cue:
```diff
@@ -75,8 +76,20 @@ ccCommand: {
+		if #flags.traffic == "external" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "internal"  {x}]
+        }
+		if #flags.traffic == "any" {
```
nit: There looks like there's something funny going on here with the spacing.

One of my TODO's is to add in `cue fmt` to help format cue files in a consistent way, I suspect that'll modify the file here.

sayboras replied:
oh nice, I think it's due to my Goland setup with tab and space. Will run cue fmt, thanks

joestringer commented on examples/kubernetes/connectivity-check/main_tool.cue:
```diff
@@ -75,8 +76,20 @@ ccCommand: {
+        }
+		if #flags.traffic == "external" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "internal"  {x}]
+        }
```
you could probably simplify this to one if condition:
```suggestion
        if #flags.traffic != "any" {
            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != #flags.traffic  {x}]
        }
```

sayboras replied:
should it be something like below as x.metadata.labels.traffic could be "any"? We also need another default case to avoid ` undefined field "resources"` issue.

```
		if #flags.traffic != "any" {
			resources: [ for x in task.filterName.resources if x.metadata.labels.traffic == #flags.traffic || x.metadata.labels.traffic == "any" {x}]
		}
		if #flags.traffic == "any" {
			resources: task.filterName.resources
		}
```

joestringer commented on examples/kubernetes/connectivity-check/network.cue:
```diff
@@ -7,8 +7,13 @@ package connectivity_check
 }
+
+_trafficCheck: {
+    metadata: labels: traffic: "external"
```
nit: it's possible to expand the _networkCheck: section to include this, for example:

```
_networkCheck: {
    metadata: labels: {
        component: "network-check"
        traffic: "external"
    }
}
```

sayboras replied:
Not sure if we can merge this together, as `pod-to-a` will have traffic as external as well.

joestringer reviewed (changes requested):
Just a few style nit comments above, but otherwise LGTM.

joestringer commented on examples/kubernetes/connectivity-check/main_tool.cue:
```diff
@@ -75,8 +76,17 @@ ccCommand: {
 
+	task: filterTraffic: {
+		if #flags.traffic != "any" {
+			resources: [ for x in task.filterName.resources if x.metadata.labels.traffic == #flags.traffic || x.metadata.labels.traffic == "any" {x}]
```
I'm not sure I understand why the "any" comparison is necessary, the line towards the top of the file marks `*"any"` as default for `#flags.traffic` via the `*`, so it should always have a value..?

joestringer replied:
^ Not a big deal though, I'm happy with the rest of this PR as-is.
